In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r'techcorp_workforce.csv',index_col='id')
df.head()

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [3]:
sales_df = pd.read_csv(r'sales.csv')

# Selecting Columns 
## As DF 
`df[['column_name']]`


In [4]:
type(df[['department']])

pandas.DataFrame

## As Series
`df['column_name']`


In [5]:
type(df['department'])

pandas.Series

## Multiple columns 
`df[['column_name1','column_name2','column_name3']]`


In [6]:
type(df[['department','salary']])

pandas.DataFrame

# Knowing Data
## How many rows and columns 
`df.shape`


In [7]:
df.shape

(51, 6)

## what are the columns 
`df.columns`


In [8]:
df.columns

Index(['first_name', 'last_name', 'department', 'salary', 'phone_number',
       'joining_date'],
      dtype='str')

## datatype of each column 
`df.dtypes`


In [9]:
df.dtypes

first_name        str
last_name         str
department        str
salary          int64
phone_number      str
joining_date      str
dtype: object

## full summary: columns, types, non-null counts
`df.info()`

In [10]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51 entries, 1 to 51
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   first_name    51 non-null     str  
 1   last_name     51 non-null     str  
 2   department    51 non-null     str  
 3   salary        51 non-null     int64
 4   phone_number  35 non-null     str  
 5   joining_date  51 non-null     str  
dtypes: int64(1), str(5)
memory usage: 2.5 KB


# Statistical summary 
## numerical datatype
`df.describe()`

In [11]:
df.describe()

,salary
count,51.000000
mean,80784.313725
std,17561.963131
min,48000.000000
25%,68500.000000
50%,81000.000000
75%,91000.000000
max,130000.000000


## all datatype 
`df.describe(include='all')`

In [12]:
df.describe(include='all')

,first_name,last_name,department,salary,phone_number,joining_date
count,51,51,51,51.000000,35,51
unique,51,50,6,NaN,35,51
top,Sarah,Mitchell,HR,NaN,555-0101,15-03-2021
freq,1,2,10,NaN,1,1
mean,NaN,NaN,NaN,80784.313725,NaN,NaN
std,NaN,NaN,NaN,17561.963131,NaN,NaN
min,NaN,NaN,NaN,48000.000000,NaN,NaN
25%,NaN,NaN,NaN,68500.000000,NaN,NaN
50%,NaN,NaN,NaN,81000.000000,NaN,NaN
75%,NaN,NaN,NaN,91000.000000,NaN,NaN


# Renaming 
## Instance column rename
`df.rename(columns={"actual name" : "new name"})`

Note:- this doesn't update the dataframe but only the instance. To save the updated instance, we can refer that to a variable.

In [13]:
renamed_df = df.rename(columns={
    "first_name" : "First Name",
    "last_name" : "Last Name",
    "phone_number" : "phone number".title(),
    "joining_date" : "joining date".title()
})
renamed_df.head()

,First Name,Last Name,department,salary,Phone Number,Joining Date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [14]:
df.rename(columns={
    "first_name" : "First Name",
    "last_name" : "Last Name",
    "phone_number" : "phone number".title(),
    "joining_date" : "joining date".title()
}).head()

,First Name,Last Name,department,salary,Phone Number,Joining Date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [15]:
df.head()

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021


## Permanent column rename
`df.rename(columns={"actual name" : "new name"},inplace=True)`

Note:- this update the dataframe and store in the same.

In [16]:
df2 = df.copy(deep=True)
df2.head()

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [17]:
df2.rename(columns={
    "first_name" : "First Name",
    "last_name" : "Last Name",
    "phone_number" : "phone number".title(),
    "joining_date" : "joining date".title()
},inplace=True)
df2.head()

,First Name,Last Name,department,salary,Phone Number,Joining Date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021


# Creating New Columns
## Simple column
New column can be created using the new name for a column that is not present already in the dataframe.
`df['new_column_name'] = df['existing_column_name']`

In [18]:
df2['new_salary']= df2['salary']
df2.head()

,First Name,Last Name,department,salary,Phone Number,Joining Date,new_salary
id,,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021,95000
2,Michael,Chen,HR,88000,555-0102,01-06-2022,88000
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021,82500
4,David,Park,HR,80000,555-0104,10-01-2023,80000
5,Lisa,Thompson,HR,65000,NaN,05-04-2021,65000


## Calculated column
any arithmetic or logical operation can  also be performed columnwise 

In [19]:
df2['new_salary_monthly']= df2['salary']/12.0
df2.head()

,First Name,Last Name,department,salary,Phone Number,Joining Date,new_salary,new_salary_monthly
id,,,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021,95000,7916.666667
2,Michael,Chen,HR,88000,555-0102,01-06-2022,88000,7333.333333
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021,82500,6875.000000
4,David,Park,HR,80000,555-0104,10-01-2023,80000,6666.666667
5,Lisa,Thompson,HR,65000,NaN,05-04-2021,65000,5416.666667


# Copy
## Refer the og
when we want to refer the original dataframe with another name 
`df1 = df`

Note:- Any change in df1 affects df too 

In [20]:
og_df = pd.read_csv(r'techcorp_workforce.csv')
og_df.head()

,id,first_name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [21]:
copy_df = og_df
copy_df.head()

,id,first_name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [22]:
copy_df.rename(columns={
    "first_name":"first name".title()
},
inplace=True)
copy_df.head()

,id,First Name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [23]:
og_df.head()

,id,First Name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


## Shallow copy 
Modifying the structure (adding/deleting columns or changing cell values via .loc) in df2 will not affect df.    
But by default in standard Python, a shallow copy can sometimes still mutate nested Python objects (like lists or dicts stored inside cells) in the original DataFrame.    
This is the standard way to safely duplicate a DataFrame for normal data analysis steps without worrying about the SettingWithCopyWarning

`df2 = df.copy()`



In [24]:
og_df = pd.read_csv(r'techcorp_workforce.csv')
og_df.head()

,id,first_name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [25]:
copy_df2 = og_df.copy()
copy_df.head()

,id,First Name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [26]:
copy_df.rename(columns={
    "first_name":"first name".title()
},
inplace=True)
copy_df.head()

,id,First Name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


In [27]:
og_df.head()

,id,first_name,last_name,department,salary,phone_number,joining_date
0,1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
1,2,Michael,Chen,HR,88000,555-0102,01-06-2022
2,3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
3,4,David,Park,HR,80000,555-0104,10-01-2023
4,5,Lisa,Thompson,HR,65000,NaN,05-04-2021


## Deep copy 
Completely isolation between both the dataframes

`df3 = df.copy(deep=True)`

in pandas by default `deep=True`

# Dupliactes
## Remove Duplicates    
Removes the rows with the duplicate values for the given column

`df[['column_name']].drop_duplicates()` 

This works on a dataframe and return a dataframe

In [28]:
orders_df = pd.read_csv(r'orders.csv')
orders_df.head() 

,id,cust_id,order_date,order_details,total_order_cost
0,1,3,04-03-2019,Coat,100
1,2,3,01-03-2019,Shoes,80
2,3,3,07-03-2019,Skirt,30
3,4,7,01-02-2019,Coat,25
4,5,7,10-03-2019,Shoes,80


In [29]:
odf1= orders_df.copy(deep=True)
unique_details_df = odf1[["order_details"]].drop_duplicates()
type(unique_details_df)

pandas.DataFrame

## Multiple column names 
When multiple columns passed, it removes rows where the entire combination is duplicated:

In [30]:
odf1[['order_details','cust_id']].drop_duplicates().head()

,order_details,cust_id
0,Coat,3
1,Shoes,3
2,Skirt,3
3,Coat,7
4,Shoes,7


## Single column
When a single column needs to be worked on we can use 

`df['column_name'].unique()`

this returns a series of unique values.

In [31]:
odf1['order_details'].unique()

<StringArray>
[   'Coat',   'Shoes',   'Skirt',   'Boats',  'Shirts', 'Slipper',   'Jeans',
    'Suit', 'Dresses',  'Gloves',     'Tie',     'Cap',  'Jacket']
Length: 13, dtype: str

### Note: 
Working on **DATA FRAME** use `drop_duplicates()`

Working on **SERIES** use `unique()`

# Filtering Rows 
Get the rows that are having values satisfying a certain criteria.

This is called as boolean indexing.

`df[df["column_name"] condition "criteria"]`


The expression inside the bracket produces a Series of True/ False values 

Conditions can be `== <= < >= > !=` 

In [32]:
df[df["department"] == "HR"]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021
6,Robert,Garcia,HR,58000,555-0106,15-11-2022
7,Jennifer,Lee,HR,72000,555-0107,20-05-2023
8,William,Davis,HR,91000,555-0108,12-08-2019
9,Patricia,Wilson,HR,78000,NaN,28-02-2018


### Internal Working

`df[df["department"] == "HR"]` <- produces a `series` of `True/ False `

This `Series` of `True` is compared with each row of the `dataframe` <- `df[boolean_series]`

The true rows are displayed. 

In [33]:
df[df[["department"]] == "HR"]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,NaN,NaN,HR,NaN,NaN,NaN
2,NaN,NaN,HR,NaN,NaN,NaN
3,NaN,NaN,HR,NaN,NaN,NaN
4,NaN,NaN,HR,NaN,NaN,NaN
5,NaN,NaN,HR,NaN,NaN,NaN
6,NaN,NaN,HR,NaN,NaN,NaN
7,NaN,NaN,HR,NaN,NaN,NaN
8,NaN,NaN,HR,NaN,NaN,NaN
9,NaN,NaN,HR,NaN,NaN,NaN


In [34]:
df[["department"]] == "HR"

,department
id,
1,True
2,True
3,True
4,True
5,True
6,True
7,True
8,True
9,True


### Internal Working

`df[df[["department"]] == "HR"]` <- produces a `dataframe` of `True/ False `

This `dataframe` creates a masking effect on the original dataframe where   
`if mask[i, j] == True:`     
`    keep value`        
`else:`          
`    set NaN`         

so pandas aligns by 
1. row index 
2. column names 

when column name matches between the 2 df (the masking and the original df) then that column gets the value. 

When the column name doesn't match `nan` is filled

In [35]:
odf1[df[["department"]]=="HR"]

,id,cust_id,order_date,order_details,total_order_cost
0,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN,NaN
4,NaN,NaN,NaN,NaN,NaN
5,NaN,NaN,NaN,NaN,NaN
6,NaN,NaN,NaN,NaN,NaN
7,NaN,NaN,NaN,NaN,NaN
8,NaN,NaN,NaN,NaN,NaN
9,NaN,NaN,NaN,NaN,NaN


## String matching 
we an use `.str` to match the strings that condition. 
 

 NOTE:-

 `.str` can be used only on the **Dataframe**

In [36]:
df[df["first_name"].str.startswith("Jo")]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
40,Jonathan,Carter,Finance,85000,555-0603,20-10-2021


In [37]:
# df[df[["first_name"]].str.endswith("th")]

# ---------------------------------------------------------------------------
# AttributeError                            Traceback (most recent call last)
# Cell In[153], line 1
# ----> 1 df[df[["first_name"]].str.endswith("th")]

# File ~\AppData\Roaming\Python\Python313\site-packages\pandas\core\generic.py:6206, in NDFrame.__getattr__(self, name)
#    6202             and name not in self._accessors
#    6203             and self._info_axis._can_hold_identifiers_and_holds_name(name)
#    6204         ):
#    6205             return self[name]
# -> 6206         return object.__getattribute__(self, name)

# AttributeError: 'DataFrame' object has no attribute 'str'

In [38]:
df[df["first_name"].str.endswith("th")]


,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
29,Kenneth,Torres,Marketing,95000,555-0405,10-03-2018


### Partial string matchings

`.contains` workswith regex patterns too. 

In [39]:
df[df["first_name"].str.contains("ha")]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
2,Michael,Chen,HR,88000,555-0102,01-06-2022
20,Stephanie,Clark,Engineering,98000,555-0304,10-03-2023
28,Sharon,Scott,Marketing,91000,555-0404,01-10-2022
40,Jonathan,Carter,Finance,85000,555-0603,20-10-2021


In [40]:
# Case-insensitive
df[df["first_name"].str.lower().str.contains("shar")]


,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
28,Sharon,Scott,Marketing,91000,555-0404,01-10-2022


# Multiple conditions

Instead of using just a single boolean filter, we can use multiple filters to filter the rows.

we use `& , |, ~` instead of `and, or, not`

This is because the former is an elementwise operator while the later is the scalar operator. 

elementwise operator compares each rows while scalar  operator compares the whole table. 

For filtering we need to compare each rows rather than the whole table. 

## And conditions

In [41]:
odf1[
    (odf1["order_details"] == "Coat") &
    (odf1["total_order_cost"] > 100)
]

,id,cust_id,order_date,order_details,total_order_cost
19,20,12,11-01-2019,Coat,125
24,25,7,19-04-2019,Coat,125


## Or conditions

In [42]:
df[
    (df.department == "HR") |
    (df.department == "Admin")
    ]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021
6,Robert,Garcia,HR,58000,555-0106,15-11-2022
7,Jennifer,Lee,HR,72000,555-0107,20-05-2023
8,William,Davis,HR,91000,555-0108,12-08-2019
9,Patricia,Wilson,HR,78000,NaN,28-02-2018


## Or on same column 
We can use `.isin()` for multiple or conditions

In [43]:
df[df["department"].isin(["HR", "Admin"])]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
4,David,Park,HR,80000,555-0104,10-01-2023
5,Lisa,Thompson,HR,65000,NaN,05-04-2021
6,Robert,Garcia,HR,58000,555-0106,15-11-2022
7,Jennifer,Lee,HR,72000,555-0107,20-05-2023
8,William,Davis,HR,91000,555-0108,12-08-2019
9,Patricia,Wilson,HR,78000,NaN,28-02-2018


## Filterning in between ranges

On the column we can use `.between(from, to)` range.

both the extremes are inclusive. 

In [44]:
odf1[odf1["total_order_cost"].between(50, 200)]


,id,cust_id,order_date,order_details,total_order_cost
0,1,3,04-03-2019,Coat,100
1,2,3,01-03-2019,Shoes,80
4,5,7,10-03-2019,Shoes,80
5,6,15,01-02-2019,Boats,100
6,7,15,11-01-2019,Shirts,60
8,9,15,01-03-2019,Jeans,80
9,10,15,09-03-2019,Shirts,50
10,11,5,01-02-2019,Shoes,80
11,12,12,11-01-2019,Shirts,60
13,14,4,01-02-2019,Shoes,80


In [45]:
odf1.loc[odf1["total_order_cost"].between(50, 200)].reset_index(drop=True)


,id,cust_id,order_date,order_details,total_order_cost
0,1,3,04-03-2019,Coat,100
1,2,3,01-03-2019,Shoes,80
2,5,7,10-03-2019,Shoes,80
3,6,15,01-02-2019,Boats,100
4,7,15,11-01-2019,Shirts,60
5,9,15,01-03-2019,Jeans,80
6,10,15,09-03-2019,Shirts,50
7,11,5,01-02-2019,Shoes,80
8,12,12,11-01-2019,Shirts,60
9,14,4,01-02-2019,Shoes,80


In [46]:
odf1.loc[odf1["total_order_cost"].between(50, 200)]

,id,cust_id,order_date,order_details,total_order_cost
0,1,3,04-03-2019,Coat,100
1,2,3,01-03-2019,Shoes,80
4,5,7,10-03-2019,Shoes,80
5,6,15,01-02-2019,Boats,100
6,7,15,11-01-2019,Shirts,60
8,9,15,01-03-2019,Jeans,80
9,10,15,09-03-2019,Shirts,50
10,11,5,01-02-2019,Shoes,80
11,12,12,11-01-2019,Shirts,60
13,14,4,01-02-2019,Shoes,80


In [47]:
odf1.loc[odf1["total_order_cost"].between(50, 200)].reset_index(drop=False)

,index,id,cust_id,order_date,order_details,total_order_cost
0,0,1,3,04-03-2019,Coat,100
1,1,2,3,01-03-2019,Shoes,80
2,4,5,7,10-03-2019,Shoes,80
3,5,6,15,01-02-2019,Boats,100
4,6,7,15,11-01-2019,Shirts,60
5,8,9,15,01-03-2019,Jeans,80
6,9,10,15,09-03-2019,Shirts,50
7,10,11,5,01-02-2019,Shoes,80
8,11,12,12,11-01-2019,Shirts,60
9,13,14,4,01-02-2019,Shoes,80


In [48]:
odf1['order_date'] = pd.to_datetime(odf1['order_date'], format='%d-%m-%Y')

odf1[
    odf1['order_date'].between("2019-04-01", "2019-04-30")
    ]

,id,cust_id,order_date,order_details,total_order_cost
15,16,3,2019-04-19,Shirts,50
16,17,7,2019-04-19,Suit,150
17,18,15,2019-04-19,Skirt,30
18,19,15,2019-04-20,Dresses,200
20,21,7,2019-04-01,Suit,50
21,22,7,2019-04-02,Skirt,30
22,23,7,2019-04-03,Dresses,50
23,24,7,2019-04-04,Coat,25
24,25,7,2019-04-19,Coat,125
25,26,3,2019-04-20,Gloves,20


## Order of execution 
1. And executes first
2. Or executes next 

`techcorp_workforce[`   
`       (techcorp_workforce["department"] == "HR") |`        
`       (techcorp_workforce["department"] == "Engineering") & `   
`       (techcorp_workforce["salary"] > 80000)`  
`]`

 What Python actually evaluates:        
 `HR (any salary) OR (Engineering AND salary > 80000)`

 so we need to use `()` to control the order

# NAN handling 

Nan - is similar to none 

this is from numpy 

`np.nan == np.nan    # False (not True!)`

`np.nan != np.nan    # True`

`np.nan > 5          # False`

`np.nan == "text"    # False`

when use any string functions the `nan` values are silently ignored as the return false 

To check the `Nan` we need to use `.isna()` or to neglect `nan` we use `.notna()`

In [49]:
df.phone_number.isna()

id
1     False
2     False
3      True
4     False
5      True
6     False
7     False
8     False
9      True
10    False
11     True
12    False
13    False
14    False
15     True
16    False
17    False
18    False
19     True
20    False
21    False
22     True
23    False
24    False
25    False
26     True
27    False
28    False
29    False
30     True
31    False
32    False
33     True
34    False
35    False
36     True
37    False
38    False
39     True
40    False
41    False
42     True
43    False
44    False
45     True
46    False
47     True
48    False
49    False
50     True
51    False
Name: phone_number, dtype: bool

In [50]:
df.isna().head()

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,False,False,False,False,False,False
2,False,False,False,False,False,False
3,False,False,False,False,True,False
4,False,False,False,False,False,False
5,False,False,False,False,True,False


In [51]:
df[df['phone_number'].isna()]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
3,Emily,Rodriguez,HR,82500,NaN,20-09-2021
5,Lisa,Thompson,HR,65000,NaN,05-04-2021
9,Patricia,Wilson,HR,78000,NaN,28-02-2018
11,Amanda,Taylor,Admin,92000,NaN,15-03-2022
15,Jessica,White,Admin,48000,NaN,14-02-2023
19,Brian,Moore,Engineering,105000,NaN,20-06-2021
22,Rebecca,Walker,Engineering,68000,NaN,01-11-2023
26,Laura,King,Marketing,85000,NaN,20-02-2022
30,Carol,Nguyen,Marketing,78000,NaN,20-11-2015


In [52]:
df[df['phone_number'].notna()]

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
4,David,Park,HR,80000,555-0104,10-01-2023
6,Robert,Garcia,HR,58000,555-0106,15-11-2022
7,Jennifer,Lee,HR,72000,555-0107,20-05-2023
8,William,Davis,HR,91000,555-0108,12-08-2019
10,James,Martinez,Admin,87000,555-0201,01-07-2021
12,Christopher,Anderson,Admin,81500,555-0203,22-08-2023
13,Michelle,Thomas,Admin,54000,555-0204,30-11-2021


## Counting Nan values 

to count the number of `nan`

`#get count in each column`

`df.isna().sum()`

In [53]:
df.isna().sum()

first_name       0
last_name        0
department       0
salary           0
phone_number    16
joining_date     0
dtype: int64

to get the percentage of data in a column having `nan`

`df.isna().mean()*100`

In [54]:
df.isna().mean()*100

first_name       0.000000
last_name        0.000000
department       0.000000
salary           0.000000
phone_number    31.372549
joining_date     0.000000
dtype: float64

## Rows with Nan
to get the number of rows with nan we use 

`df.isna().any(axis=1).sum()`

In [55]:
df.isna().any(axis=1)

id
1     False
2     False
3      True
4     False
5      True
6     False
7     False
8     False
9      True
10    False
11     True
12    False
13    False
14    False
15     True
16    False
17    False
18    False
19     True
20    False
21    False
22     True
23    False
24    False
25    False
26     True
27    False
28    False
29    False
30     True
31    False
32    False
33     True
34    False
35    False
36     True
37    False
38    False
39     True
40    False
41    False
42     True
43    False
44    False
45     True
46    False
47     True
48    False
49    False
50     True
51    False
dtype: bool

In [56]:
df.isna().any(axis=1).sum()

np.int64(16)

In [57]:
df.isna().any(axis=1).count()

np.int64(51)

### Differnce in the output 

`.sum()` adds the values <- in python `True = 1` and `False = 0`

but `.count()` counts the non-null vales <- in pthon 0 and  1 are not non-null 

## Columns with NAN

In [58]:
df.isna().any(axis=0)

first_name      False
last_name       False
department      False
salary          False
phone_number     True
joining_date    False
dtype: bool

In [59]:
df.isna().any(axis=0).sum()

np.int64(1)

## Fill the `nan` values

Use `.fillna()` to fill the null values, it takes one argument that replaces `nan`

In [60]:
df.phone_number.fillna('Unkown')

id
1     555-0101
2     555-0102
3       Unkown
4     555-0104
5       Unkown
6     555-0106
7     555-0107
8     555-0108
9       Unkown
10    555-0201
11      Unkown
12    555-0203
13    555-0204
14    555-0205
15      Unkown
16    555-0207
17    555-0301
18    555-0302
19      Unkown
20    555-0304
21    555-0305
22      Unkown
23    555-0307
24    555-0308
25    555-0401
26      Unkown
27    555-0403
28    555-0404
29    555-0405
30      Unkown
31    555-0501
32    555-0502
33      Unkown
34    555-0504
35    555-0505
36      Unkown
37    555-0507
38    555-0601
39      Unkown
40    555-0603
41    555-0604
42      Unkown
43    555-0606
44    555-0607
45      Unkown
46    555-0109
47      Unkown
48    555-0309
49    555-0310
50      Unkown
51    555-0608
Name: phone_number, dtype: str

## Drop the `nan` values 

we can dropthe based on 2 conditions
1. any column has `nan`   <- `.dropna()`>
2. specific column has `nan`  <- `.dropna(subset='column-name')`

we use `.dropna()` <- for dropping 




In [61]:
# Drop rows where ANY column is NaN
df.dropna().head()

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
4,David,Park,HR,80000,555-0104,10-01-2023
6,Robert,Garcia,HR,58000,555-0106,15-11-2022
7,Jennifer,Lee,HR,72000,555-0107,20-05-2023


Drop na does not alter the existing table.. we need to store the result in another variable

In [62]:
# Drop rows where a SPECIFIC column is NaN
df.dropna(subset=["phone_number"]).head()

,first_name,last_name,department,salary,phone_number,joining_date
id,,,,,,
1,Sarah,Mitchell,HR,95000,555-0101,15-03-2021
2,Michael,Chen,HR,88000,555-0102,01-06-2022
4,David,Park,HR,80000,555-0104,10-01-2023
6,Robert,Garcia,HR,58000,555-0106,15-11-2022
7,Jennifer,Lee,HR,72000,555-0107,20-05-2023


In [63]:
df.isna().sum()

first_name       0
last_name        0
department       0
salary           0
phone_number    16
joining_date     0
dtype: int64

# Sorting data

Sorting  database based on a row 

`.sort_values()` <- ascending          
`.sort_values(ascending =False)` <- descending


In [64]:
# odf1.sort_values()

# ---------------------------------------------------------------------------
# TypeError                                 Traceback (most recent call last)
# Cell In[83], line 1
# ----> 1 odf1.sort_values()

# TypeError: DataFrame.sort_values() missing 1 required positional argument: 'by'

In [65]:
odf1.sort_values('total_order_cost').head()

,id,cust_id,order_date,order_details,total_order_cost
27,28,3,2019-04-22,Cap,15
7,8,15,2019-03-11,Slipper,20
25,26,3,2019-04-20,Gloves,20
12,13,12,2019-03-11,Slipper,20
26,27,3,2019-04-21,Tie,25


In [66]:
odf1.sort_values(['total_order_cost','cust_id']).head()

,id,cust_id,order_date,order_details,total_order_cost
27,28,3,2019-04-22,Cap,15
25,26,3,2019-04-20,Gloves,20
12,13,12,2019-03-11,Slipper,20
7,8,15,2019-03-11,Slipper,20
26,27,3,2019-04-21,Tie,25


In [67]:
sorted(odf1)

['cust_id', 'id', 'order_date', 'order_details', 'total_order_cost']

### Differnce in output 

`.sort_values()` <- pandas function, takes the df and works row by row and returns dataframe op

`sorted()` <- python based, takes iterable. for a dataframe the elements are its column this when iterated the value is the column names sorted. 

In [68]:
odf1.head()

,id,cust_id,order_date,order_details,total_order_cost
0,1,3,2019-03-04,Coat,100
1,2,3,2019-03-01,Shoes,80
2,3,3,2019-03-07,Skirt,30
3,4,7,2019-02-01,Coat,25
4,5,7,2019-03-10,Shoes,80


## Creating a new column 
When calculation need & requirements are different then we might need to create a new column for find the values and then sorting them 

this can be avoided if the purpose is only for sorting using lambda 

for example:
`df.sort_values(key=lambda c: c.str.len())`


In [69]:
df["name_len"] = df["first_name"].str.len()
df.sort_values("name_len",ascending=False).head()

,first_name,last_name,department,salary,phone_number,joining_date,name_len
id,,,,,,,
12,Christopher,Anderson,Admin,81500,555-0203,22-08-2023,11
20,Stephanie,Clark,Engineering,98000,555-0304,10-03-2023,9
7,Jennifer,Lee,HR,72000,555-0107,20-05-2023,8
9,Patricia,Wilson,HR,78000,NaN,28-02-2018,8
40,Jonathan,Carter,Finance,85000,555-0603,20-10-2021,8


In [70]:
for x in copy_df["First Name"]:
    print(x)

Sarah
Michael
Emily
David
Lisa
Robert
Jennifer
William
Patricia
James
Amanda
Christopher
Michelle
Daniel
Jessica
Matthew
Kevin
Angela
Brian
Stephanie
Timothy
Rebecca
George
Donna
Ryan
Laura
Eric
Sharon
Kenneth
Carol
Jason
Amy
Jacob
Kathleen
Gary
Anna
Stephen
Nicholas
Brenda
Jonathan
Shirley
Edward
Cynthia
Jeffrey
Dorothy
Mark
Sandra
Paul
Kimberly
Andrew
Betty


In [71]:
copy_df.sort_values(by="First Name", key=lambda c: c.str.len(), ascending=False).head()

,id,First Name,last_name,department,salary,phone_number,joining_date
11,12,Christopher,Anderson,Admin,81500,555-0203,22-08-2023
19,20,Stephanie,Clark,Engineering,98000,555-0304,10-03-2023
6,7,Jennifer,Lee,HR,72000,555-0107,20-05-2023
8,9,Patricia,Wilson,HR,78000,NaN,28-02-2018
39,40,Jonathan,Carter,Finance,85000,555-0603,20-10-2021


# Limiting Data 

## with `.sort_values()`
 
We can limit using 

`.head()` <- top values 

`.tail()` <- bottom values

Useful for multiple column sorting or specific sort order

## without `.sort_values()`

without `sort_values()` we can use 

`.nsmallest(N, "sorting column")` <- for ascending 

`.nlargest(N, "sorting column")` <- for descending

Usualy faster than the other method

In [72]:
df.nlargest(5, "salary")

,first_name,last_name,department,salary,phone_number,joining_date,name_len
id,,,,,,,
23,George,Hall,Engineering,130000,555-0307,01-09-2016,6
17,Kevin,Brown,Engineering,125000,555-0301,15-01-2021,5
18,Angela,Miller,Engineering,118000,555-0302,01-04-2022,6
19,Brian,Moore,Engineering,105000,NaN,20-06-2021,5
44,Jeffrey,Evans,Finance,105000,555-0607,01-06-2014,7


In [73]:
df.nsmallest(3, "salary")

,first_name,last_name,department,salary,phone_number,joining_date,name_len
id,,,,,,,
15,Jessica,White,Admin,48000,NaN,14-02-2023,7
47,Sandra,Parker,Admin,51000,NaN,01-02-2024,6
13,Michelle,Thomas,Admin,54000,555-0204,30-11-2021,8


# Full Pattern 
In practice, you almost never sort alone. 

The real pattern is a chain: 
1. filter to the rows you care about, 
2. sort by the column that matters, 
3. limit to the top N, and 
4. select only the columns you need. 

This chain is the backbone of exploratory analysis.

In [74]:
df[df["department"] == "Engineering"].sort_values("joining_date", ascending=False).head(5)[["first_name", "last_name", "joining_date"]]

,first_name,last_name,joining_date
id,,,
19,Brian,Moore,20-06-2021
24,Donna,Allen,15-12-2019
21,Timothy,Lewis,15-08-2022
17,Kevin,Brown,15-01-2021
48,Paul,Collins,10-03-2024


# Aggregate functions 

## number of rows 

`len(dataframe)`

In [75]:
len(odf1)

31

## no. non null values per column 

`df.count()`

In [76]:
odf1.count()

id                  31
cust_id             31
order_date          31
order_details       31
total_order_cost    31
dtype: int64

## unique no. of non-null values per column

`df.nunique()`

In [77]:
odf1.nunique()

id                  31
cust_id              7
order_date          17
order_details       13
total_order_cost    12
dtype: int64

In [78]:
odf1.cust_id.nunique()

7

### Understanding 
`.unique()` works on **series** <- unique values in a particular column

`.nunique()` works on both **series and dataframe** <- unique values in each column of a particular 

## numerical column analysis

### sum of specified column 


`df[ _numerical_column_ ].sum()`

`.sum()` ignores `nan` values 

In [79]:
odf1.total_order_cost.sum()

np.int64(2215)

### Mean of the column

`df[ _numerical_column_ ].mean()`

`.mean()` skips `nan` values. 

hence the rows with `nan` values will not be inculed in the mean calculation 

therefore, 
 If you have values 100, 200, and NaN, the mean is 150 (300 / 2), not 100 (300 / 3)

In [80]:
odf1.total_order_cost.mean()

np.float64(71.45161290322581)

## Multiple functions at once 

to use `sum, max, min, mean, count, etc`  we can use `.agg([ _list_of_functions_ ])`

`.agg()` is a function that takes a **Series** and returns a **scalar OR Series**

In [81]:
odf1["total_order_cost"].agg(["sum", "mean", "min", "max","count"])

sum      2215.000000
mean       71.451613
min        15.000000
max       200.000000
count      31.000000
Name: total_order_cost, dtype: float64

In [82]:
df.columns

Index(['first_name', 'last_name', 'department', 'salary', 'phone_number',
       'joining_date', 'name_len'],
      dtype='str')

In [83]:
df['salary'].agg(['min', 'mean', 'max']).rename(
    index={'min': 'lowest_salary', 'mean': 'mean_salary', 'max': 'highest_salary'}
).to_frame().T

,lowest_salary,mean_salary,highest_salary
salary,48000.0,80784.313725,130000.0


In [84]:
sales_df[sales_df['salesperson'].str.lower().isin(["samantha","lisa"])]['sales_revenue'].sum()

np.int64(112650)

In [85]:
df[df['salary']==df['salary'].max()]['first_name']

id
23    George
Name: first_name, dtype: str

We can add custom function inside `.agg()` function 

In [ ]:
print(df['salary'].agg(['mean', 'max', lambda x: x.min()]))
print("Normal:\n",df['salary'].agg(['mean', 'max', 'min']))
print("Custom:\n",df['salary'].agg(['mean', 'max', lambda x: x.max() - x.min()]))


mean         80784.313725
max         130000.000000
<lambda>     48000.000000
Name: salary, dtype: float64
Normal:
 mean     80784.313725
max     130000.000000
min      48000.000000
Name: salary, dtype: float64
Custom:
 mean         80784.313725
max         130000.000000
<lambda>     82000.000000
Name: salary, dtype: float64
